src/sage/data_structures/pairing_heap.pxd

```diff
@@ -0,0 +1,82 @@
+# ******************************************************************************
+# Copyright (C) 2024 David Coudert <david.coudert@inria.fr>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ******************************************************************************
+
+# ==============================================================================
+# Interface to pairing heap data structure from ./pairing_heap.h
+# ==============================================================================
+
+from libcpp.pair cimport pair
+
+cdef extern from "./pairing_heap.h" namespace "pairing_heap":
+    cdef cppclass PairingHeap[TypeOfItem, TypeOfValue]:
+        PairingHeap() except +
+        PairingHeap(PairingHeap[TypeOfItem, TypeOfValue]) except +
+        bint empty()
+        void reset()
+        void push(TypeOfItem, TypeOfValue) except +
+        pair[TypeOfItem, TypeOfValue] top() except +
+        TypeOfItem top_item() except +
+        TypeOfValue top_value() except +
+        void pop() except +
+        void decrease(TypeOfItem, TypeOfValue) except +
+        bint contains(TypeOfItem)
+        TypeOfValue value(TypeOfItem) except +
+
+# ==============================================================================
+# Pairing heap data structure with fixed capacity n
+# ==============================================================================
+
+from sage.data_structures.bitset_base cimport bitset_t
+
+ctypedef struct PairingHeapNode:
+    void * value             # value associated with the item
+    PairingHeapNode * prev   # Previous sibling of the node or parent
+    PairingHeapNode * succ   # Next sibling of the node
+    PairingHeapNode * child  # First child of the node
+
+
+cdef bint _compare(PairingHeapNode * a, PairingHeapNode * b) except *
+cdef PairingHeapNode * _pair(PairingHeapNode * p) except *
+cdef PairingHeapNode * _merge(PairingHeapNode * a, PairingHeapNode * b) except *
+cdef _link(PairingHeapNode * a, PairingHeapNode * b) except *
+cdef _unlink(PairingHeapNode * p) except *
+
+
+cdef class PairingHeap_class:
+    cdef str name                 # name of the data structure
```

Comment:
The string seems to store the name of the class, could it be a static attribute to avoid storing one string in each class object ?

Reply:
actually it suffices to check if the class is an instance of `PairingHeap_of_n_integers`.